mapbox/vtvalidate
# Web Assembly support

This adds Web Assembly support. This is an experiment on top of #14.

What this includes is:

 - A build script to compile the code to .wasm
 - A demo HTML page that confirms that the binary works in the browser.
 - A demo CLI that validates a vector tile (like https://github.com/mapbox/vtvalidate/pull/2 but using the wasm binary)
 - A `compare.js` benchmark comparing the speed of validation in .wasm vs platform-specific native code.

To test the code in the browser do:

```bash
./build-wasm.sh
cd html
python -m SimpleHTTPServer
```

Then view http://localhost:8000/ and drag a file onto the browser like https://github.com/mapbox/mvt-fixtures/blob/master/fixtures/003/tile.mvt. It should report `fail unknown geometry type`.

To test the code via node on the command line do:

```
cd html
node compare.js
```

My results locally are:

```
$ node compare.js 
node: 51.436ms
wasm: 512.087ms
```

Which indicates, preliminarily, that wasm is 10x slower than native code.

Note, currently the build script is failing on linux with:

```
/home/travis/build/mapbox/vtvalidate/wasm-toolchain/emsdk-portable/clang/e1.37.33_64bit/llc: relocation error: /home/travis/build/mapbox/vtvalidate/wasm-toolchain/emsdk-portable/clang/e1.37.33_64bit/llc: symbol _ZNKSt3_V214error_category10_M_messageB5cxx11Ei, version GLIBCXX_3.4.21 not defined in file libstdc++.so.6 with link time reference
```

Still need to work out why.
